### DISCUSSION TITLE
Libssh2 connects to a remote server

### DISCUSSION THREAD

**Discussion by fightingLKH** [libssh2/libssh2#1489] [Q&A]
#include <libssh2.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>

int main() {
    const char *hostname = "server_ip";
    const char *username = "username";
    const char *password = "password";
    const int port = 22;

    libssh2_init(0);
    LIBSSH2_SESSION *session = libssh2_session_init();
    
    int sock = socket(AF_INET, SOCK_STREAM, 0);
    struct sockaddr_in sin;
    sin.sin_family = AF_INET;
    sin.sin_port = htons(port);
    inet_pton(AF_INET, hostname, &sin.sin_addr);

    if (connect(sock, (struct sockaddr*)(&sin), sizeof(struct sockaddr_in)) != 0) {
        perror("connect failed");
        return 1;
    }

    if (libssh2_session_handshake(session, sock)) {
        fprintf(stderr, "Error establishing SSH session\n");
        return 1;
    }

    if (libssh2_userauth_password(session, username, password)) {
        const char *error_msg = libssh2_session_last_error(session, NULL, NULL, 0);
        fprintf(stderr, "Authentication failed: %s\n", error_msg);
        return 1;
    }

    LIBSSH2_CHANNEL *channel = libssh2_channel_open_session(session);
    if (!channel) {
        fprintf(stderr, "Unable to open channel\n");
        return 1;
    }

    if (libssh2_channel_exec(channel, "ls")) {
        fprintf(stderr, "Unable to execute command\n");
    } else {
        char buffer[1024];
        int bytes_read;

        // 读取标准输出
        while ((bytes_read = libssh2_channel_read(channel, buffer, sizeof(buffer))) > 0) {
            fwrite(buffer, 1, bytes_read, stdout);
        }

    }

    libssh2_channel_close(channel);
    libssh2_channel_free(channel);
    libssh2_session_disconnect(session, "Goodbye");
    libssh2_session_free(session);
    libssh2_exit();

    return 0;
}

I executed the above code to connect to a remote server through libssh2, invented the command ls, and read the result. After adding the print information, I found that after the function libssh2_channel_exe () was executed, the program was constantly blocked, and no information was printed, no execution was continued, and no errors were reported. May I ask what the reason is and what solutions are available?

**Comment by zszijtx** [comment]
try use libssh2_session_set_blocking(session, 1) set session block
